source/ddbus/thin.d

```diff
@@ -463,92 +503,88 @@ struct DBusAny {
       }
     } else static if (is(T == DBusAny)) {
       return this;
-    } else static if (isIntegral!T || isFloatingPoint!T) {
-      switch (type) {
-      case typeCode!byte:
-        return cast(T) int8;
-      case typeCode!short:
-        return cast(T) int16;
-      case typeCode!ushort:
-        return cast(T) uint16;
-      case typeCode!int:
-        return cast(T) int32;
-      case typeCode!uint:
-        return cast(T) uint32;
-      case typeCode!long:
-        return cast(T) int64;
-      case typeCode!ulong:
-        return cast(T) uint64;
-      case typeCode!double:
-        return cast(T) float64;
-      default:
-        throw new Exception("Can't convert type " ~ cast(char) type ~ " to " ~ T.stringof);
-      }
-    } else static if (is(T == bool)) {
-      if (type == 'b') {
-        return boolean;
-      } else {
-        throw new Exception("Can't convert type " ~ cast(char) type ~ " to " ~ T.stringof);
-      }
-    } else static if (isSomeString!T) {
-      if (type == 's') {
-        return str.to!T;
-      } else if (type == 'o') {
-        return obj.toString();
-      } else {
-        throw new Exception("Can't convert type " ~ cast(char) type ~ " to " ~ T.stringof);
-      }
-    } else static if (is(T == ObjectPath)) {
-      if (type == 'o') {
-        return obj;
-      } else {
-        throw new Exception("Can't convert type " ~ cast(char) type ~ " to " ~ T.stringof);
-      }
-    } else static if (isDynamicArray!T) {
-      if (type != 'a' && type != 'r') {
-        throw new Exception("Can't convert type " ~ cast(char) type ~ " to an array");
-      }
+    } else {
+      // In here are all static if blocks that may fall through to the throw
+      // statement at the bottom of this block.
 
-      T ret;
-      if (signature == ['y']) {
-        static if (isIntegral!(ElementType!T)) {
-          foreach (elem; binaryData) {
-            ret ~= elem.to!(ElementType!T);
+      static if (is(T == DictionaryEntry!(K, V), K, V)) {
+        if (type == 'e') {
+          static if (is(T == typeof(entry))) {
+            return entry;
+          } else {
+            return DictionaryEntry(entry.key.to!K, entry.value.to!V);
           }
         }
-      } else {
-        foreach (elem; array) {
-          ret ~= elem.to!(ElementType!T);
+      } else static if (isAssociativeArray!T) {
+        if (type == 'a' && (!array.length || array[0].type == 'e')) {
+          Unqual!(ValueType!T)[Unqual!(KeyType!T)] ret;
```

Comment:
this could be written a bit nicer. For example 2 previous lines `alias KeyT = KeyType!T; alias ValueT = ValueType!T; ValueT[KeyT] ret;`